File: src/main/resources/public/ts/services/ResponseService.ts
import {idiom, ng, notify, moment} from 'entcore';
import http from 'axios';
import {Question, Response, Types} from "../models";
import {DataUtils} from "../utils/data";

export interface ResponseService {
    list(question: Question, nbLines: number) : Promise<any>;
    countByQuestion(questionId : number) : Promise<any>;
    listMineByDistribution(questionId: number, distributionId: number) : Promise<any>;
    listByDistribution(distributionId: number) : Promise<any>;
    get(responseId: number) : Promise<any>;
    save(response: Response, questionType?: number) : Promise<any>;
    create(response: Response) : Promise<any>;
    fillResponses(formId: number, distributionId: number) : Promise<any>;
    update(response: Response) : Promise<any>;
    delete(responseId: number) : Promise<any>;
}

export const responseService: ResponseService = {
    async list(question: Question, nbLines: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/questions/${question.id}/responses?nbLines=${nbLines}&formId=${question.form_id}`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.list'));
            throw err;
        }
    },
    async countByQuestion (questionId:number):Promise<any>{ //count
        try{
            return DataUtils.getData(await http.get(`/formulaire/questions/${questionId}/responses/count`));
        }catch(err){
            notify.error(idiom.translate('formulaire.error.responseService.list'));
            throw err;
        }
    },

    async listMineByDistribution(questionId: number, distributionId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/questions/${questionId}/responses/${distributionId}`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.list'));
            throw err;
        }
    },

    async listByDistribution(distributionId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/responses/distrib/${distributionId}`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.list'));
            throw err;
        }
    },

    async get(responseId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/responses/${responseId}`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.get'));
            throw err;
        }
    },

    async save(response: Response, questionType?: number) : Promise<any> {
        if (!response.answer) {
            response.answer = "";
        }
        else {
            if (questionType === Types.TIME) {
                if (typeof response.answer != "string") {
                    response.answer = moment(response.answer).format("HH:mm");
                }
            }
            else if (questionType === Types.DATE) {
                if (typeof response.answer != "string") {
                    response.answer = moment(response.answer).format("DD/MM/YYYY");
                }
            }
        }
        return response.id ? await this.update(response) : await this.create(response);
    },

    async create(response: Response) : Promise<any> {
        try {
            return DataUtils.getData(await http.post(`/formulaire/questions/${response.question_id}/responses`, response));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.create'));
            throw err;
        }
    },

    async fillResponses(formId: number, distributionId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.post(`/formulaire/forms/${formId}/responses/fill/${distributionId}`, {}));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.create'));
            throw err;
        }
    },

    async update(response: Response) : Promise<any> {
        try {
            return DataUtils.getData(await http.put(`/formulaire/responses/${response.id}`, response));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.update'));
            throw err;
        }
    },

    async delete(responseId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.delete(`/formulaire/responses/${responseId}`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.delete'));
            throw err;
        }
    }
};

export const ResponseService = ng.service('ResponseService', (): ResponseService => responseService);

File: src/main/java/fr/openent/formulaire/service/ResponseService.java
package fr.openent.formulaire.service;

import fr.wseduc.webutils.Either;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import org.entcore.common.user.UserInfos;

public interface ResponseService {
    /**
     * List all the responses to a specific question
     * @param questionId question identifier
     * @param nbLines number of lines already loaded
     * @param handler function handler returning JsonArray data
     */
    void list(String questionId, String nbLines, JsonArray distribs, Handler<Either<String, JsonArray>> handler);

    /**
     * List all my responses to a specific question for a specific distribution
     * @param questionId question identifier
     * @param distributionId distribution identifier
     * @param user user connected
     * @param handler function handler returning JsonArray data
     */
    void listMineByDistribution(String questionId, String distributionId, UserInfos user, Handler<Either<String, JsonArray>> handler);

    /**
     * List all responses for a specific distribution
     * @param distributionId distribution identifier
     * @param handler function handler returning JsonArray data
     */
    void listByDistribution(String distributionId, Handler<Either<String, JsonArray>> handler);

    /**
     * List all responses for a specific form
     * @param formId form identifier
     * @param handler function handler returning JsonArray data
     */
    void listByForm(String formId, Handler<Either<String, JsonArray>> handler);

    /**
     * Count all the responses to a list of questions
     * @param questionIds questions identifiers
     * @param handler function handler returning JsonObject data
     */
    void countByQuestions(JsonArray questionIds, Handler<Either<String, JsonObject>> handler);

    /**
     * Get a specific response by id
     * @param responseId response identifier
     * @param handler function handler returning JsonObject data
     */
    void get(String responseId, Handler<Either<String, JsonObject>> handler);

    /**
     * Get question ids of a specific form where specific distribution has missing responses
     * @param formId form identifier
     * @param distributionId distribution identifier
     * @param handler function handler returning JsonArray data
     */
    void getMissingResponses(String formId, String distributionId, Handler<Either<String, JsonArray>> handler);

    /**
     * Create a response
     * @param response JsonObject data
     * @param user user connected
     * @param questionId question identifier
     * @param handler function handler returning JsonObject data
     */
    void create(JsonObject response, UserInfos user, String questionId, Handler<Either<String, JsonObject>> handler);

    /**
     * Create empty responses to fill missing ones for a specific distribution
     * @param questionIds JsonArray ids of missing questions
     * @param distributionId distribution identifier
     * @param user user connected
     * @param handler function handler returning JsonObject data
     */
    void fillResponses(JsonArray questionIds, String distributionId, UserInfos user, Handler<Either<String, JsonObject>> handler);

    /**
     * Update a specific response
     * @param user user connected
     * @param responseId response identifier
     * @param response JsonObject data
     * @param handler function handler returning JsonObject data
     */
    void update(UserInfos user, String responseId, JsonObject response, Handler<Either<String, JsonObject>> handler);

    /**
     * Delete specific responses
     * @param responseIds responses identifiers
     * @param handler function handler returning JsonArray data
     */
    void delete(JsonArray responseIds, Handler<Either<String, JsonArray>> handler);

    /**
     * Delete responses by distribution id
     * @param String distribution identifier
     * @param handler function handler returning JsonArray data
     */
    void deleteMultipleByDistribution(String String, Handler<Either<String, JsonArray>> handler);

    /**
     * Get all the responders of a specific form to export into a CSV file
     * @param formId form identifier
     * @param handler function handler returning JsonArray data
     */
    void getExportCSVResponders(String formId, Handler<Either<String, JsonArray>> handler);

    /**
     * Get all the responses of a specific form to export into a CSV file
     * @param formId form identifier
     * @param handler function handler returning JsonArray data
     */
    void exportCSVResponses(String formId, Handler<Either<String, JsonArray>> handler);

    /**
     * Get all the responses of a specific form to export into a PDF file
     * @param formId form identifier
     * @param handler function handler returning JsonArray data
     */
    void exportPDFResponses(String formId, Handler<Either<String, JsonArray>> handler);

    /**
     * Delete responses too old according to lifetime RGPD chosen
     * @param distributionIds id of deleted distributions
     * @param handler function handler returning JsonObject data
     */
    void deleteOldResponse(JsonArray distributionIds, Handler<Either<String, JsonArray>> handler);
}